codegen/src/pushable.rs

```diff
@@ -0,0 +1,179 @@
+        let field_idents2 = &field_idents;
+        let variant_ident = &variant.ident;
+
+        let pattern = match &variant.fields {
+            Fields::Named(_) => quote! { #ident::#variant_ident{ #(#field_idents2),* } },
+            Fields::Unnamed(_) => quote! { #ident::#variant_ident( #(#field_idents2),* ) },
+            Fields::Unit => quote! { #ident::#variant_ident },
+        };
+
+        let push_impl = gen_push_impl(tag, &field_idents, &field_types);
+
+        quote! {
+            #pattern => {
+                #push_impl
+            }
+        }
+    });
+
+    // build the final match from the individual arms
+    let push_impl = quote! {
+        match self {
+            #(#match_arms),*
+        };
+    };
+
+    gen_impl(&ident, generics, push_impl)
+}
+
+fn gen_impl(ident: &Ident, generics: Generics, push_impl: TokenStream) -> TokenStream {
+    // add trailing comma or add a where if there are no predicates
+    let (_, ty_generics, where_clause) = generics.split_for_impl();
+    let where_clause = where_clause
+        .map(|clause| quote! { #clause, })
+        .unwrap_or(quote! { where });
+
+    // generate the generic params for the impl block
+    // these need an additional lifetime used by Pushable
+    let mut generics = generics.clone();
+    let lifetime = GenericParam::from(LifetimeDef::new(Lifetime::new("'__vm", Span::call_site())));
+    generics.params.insert(0, lifetime);
+    let (impl_generics, ..) = generics.split_for_impl();
+
+    let pushable_bounds = create_pushable_bounds(&generics);
+
+    quote! {
+        #[automatically_derived]
+        #[allow(unused_attributes)]
+        impl #impl_generics ::gluon::vm::api::Pushable<'__vm> for #ident #ty_generics
+        #where_clause #(#pushable_bounds),*
+        {
+            fn push(self, vm: &'__vm ::gluon::vm::thread::Thread, ctx: &mut ::gluon::vm::thread::Context) -> ::gluon::vm::Result<()> {
+                #push_impl
+                Ok(())
+            }
+        }
+    }
+}
+
+fn gen_push_impl(tag: usize, field_idents: &[Cow<Ident>], field_types: &[&Type]) -> TokenStream {
+    debug_assert!(field_idents.len() == field_types.len());
+    let len = field_idents.len();
+
+    // push each field onto the stack
+    // this has to be done in reverse order so the fields come out in the correct
+    // order when popping the stack
+    let stack_pushes = field_idents
+        .iter()
+        .zip(field_types)
+        .map(|(ident, ty)| {
+            quote! {
+                <#ty as ::gluon::vm::api::Pushable<'__vm>>::push(#ident, vm, ctx)?;
+            }
+        })
+        .rev();
+
+    // since the number of fields is statically known, we can allocate all values in an array
+    let array_init = iter::repeat(quote! { ::gluon::vm::internal::Value::int(0) }).take(len);
+    let stack_init = quote! {
+        let mut fields = [ #(#array_init),* ];
+
+        for i in 0..#len {
+            fields[i] = ctx.stack.pop();
+        }
```

Comment:
Could generate this as just
```rust
let array_init = iter::repeat(quote! { ctx.stack.pop() }).take(len);
let stack_init = quote! {
        let fields = [ #(#array_init),* ];
```

and avoid the initialization step altogether

Reply:
Yep, that's a lot cleaner and doesn't rely on the compiler optimizing things correctly. Already fixed it with https://github.com/gluon-lang/gluon/pull/541/commits/90a428cd6977511ba7687a9e261023aaf14b57ed :)